### ISSUE TITLE
Macros like 'major' still leak, or are protected too aggressively

### ISSUE THREAD

**Issue by eddelbuettel** [RcppCore/Rcpp#272]
We had thought #227 was taken care of, but apparently not. 

Fuller discussion is on [rcpp-devel](http://thread.gmane.org/gmane.comp.lang.r.rcpp/8091) with full example leading to

```
RcppExports.cpp: In function ‘SEXPREC* majortest_rcpp_hello_world(SEXP)’:
RcppExports.cpp:15:44: error: ‘major’ was not declared in this scope
     __result = Rcpp::wrap(rcpp_hello_world(major));
```

so it looks like `major` may be undefined even when it wasn't a macro?

@kevinushey : do you spare cycles to tackle this, again?  Or do we just shake our head and call it a documentation issue about symbols to be avoided?

**Comment by jackwasey** [comment]
Simply writing the function which demonstrated #227 isn't enough to trigger the error:
```
cppFunction("int foo(int major) { return(2*major); }")
```
does compile with the addition of -std=c++11

Steps to reproduce. Firstly, in R:
```
library(Rcpp)
Rcpp.package.skeleton(attributes = TRUE, example_code = TRUE,
name="majortest")
```

Create majortest/src/Makevars containing:

```
PKG_CXXFLAGS = $(CXX1XSTD)
```

Edit rcpp_hello_world.cpp replacing:
```
List rcpp_hello_world() {
```
with
```
List rcpp_hello_world(CharacterVector major) {
```

Now both `devtools::load_all()` or `R CMD INSTALL majortest` then produce
the previously described compilation errors, with my compiler set to gcc
or clang. I am using clang 3.7 and gcc 4.9.1 on 64 bit ubuntu 14.10.

Here is the clang compiler error:
```
clang++ -I/usr/share/R/include -DNDEBUG
-I"/usr/local/lib/R/site-library/Rcpp/include"  -std=c++11 -fpic  -g -O2
-fstack-protector-strong -Wformat -Werror=format-security
-D_FORTIFY_SOURCE=2 -g  -c RcppExports.cpp -o RcppExports.o
RcppExports.cpp:15:44: error: use of undeclared identifier 'major'
    __result = Rcpp::wrap(rcpp_hello_world(major));
```
Here is the generated code from RcppExports.cpp:
```
#include <Rcpp.h>

using namespace Rcpp;

// rcpp_hello_world
List rcpp_hello_world(CharacterVector major);
RcppExport SEXP majmin11_rcpp_hello_world(SEXP majorSEXP) {
BEGIN_RCPP
    Rcpp::RObject __result;
    Rcpp::RNGScope __rngScope;
    Rcpp::traits::input_parameter< CharacterVector >::type major(majorSEXP);
    __result = Rcpp::wrap(rcpp_hello_world(major));
    return __result;
END_RCPP
}
```

**Comment by kevinushey** [comment]
Thanks -- here's a reproducible script.

```sh
#!/usr/bin/env sh

mkdir ~/scratch
cd ~/scratch
rm -rf majortest
R --vanilla --slave -e "Rcpp::Rcpp.package.skeleton(attributes = TRUE, example_code = TRUE, name = 'majortest')"
cd majortest
echo "PKG_CXXFLAGS = \$(CXX1XSTD)" > src/Makevars
sed -i "s/rcpp_hello_world()/rcpp_hello_world(CharacterVector major)/g" src/rcpp_hello_world.cpp
R --vanilla --slave -e "Rcpp::compileAttributes()"
R CMD INSTALL .
```
Will take another look.

**Comment by kevinushey** [comment]
Issue is basically that the majors leak earlier than the R includes when `-std=c++11` is specified. Have a simple fix coming.

**Comment by coatless** [comment]
This issue was addressed in PR #275.

Please close. (Tagged #506)
